config_context renders against incorrect pack

jinpingh opened the pull request:
Fixes https://github.com/StackStorm/st2/issues/4567 config_context renders against incorrect pack
Root cause:
When a action calls other action which belongs to the different pack, during `render_live_params`, only parent pack configuration is retrieved.

Fixes:
Passes action reference to `render_live_params`, if action parent pack is different with child action pack, the configuration for child pack is retired and rendered.

jinpingh commented:
> If the pack name being passed into the render_live_params is not correct, why not assign the correct pack at https://github.com/StackStorm/st2/blob/master/st2common/st2common/services/workflows.py#L546?

Running task is belong to the `pack1` and it is passed. Action execution is belong to the different pack, e.g. `pack2`, in case both pack information needed on `pack2` action.

jinpingh commented:
> If the pack name being passed into the render_live_params is not correct, why not assign the correct pack at https://github.com/StackStorm/st2/blob/master/st2common/st2common/services/workflows.py#L546?

I am thinking if `pack2` action needs both `pack1` and `pack2` information, it is good to render both.

m4dcoder commented:
What's the expected behavior from our current docs?

jinpingh commented:
> What's the expected behavior from our current docs?

Let me google.  If it is not supported, will change code to pass right pack name

jinpingh commented:
https://github.com/StackStorm/st2/pull/4570#issuecomment-468407791
Googed Stackstorm documentation and did not see anywhere mentioned that parent pack is supported in the child action execution.

m4dcoder commented on st2common/st2common/services/workflows.py:
```diff
@@ -357,7 +357,7 @@ def request_resume(ac_ex_db):
 def request_cancellation(ac_ex_db):
     wf_ac_ex_id = str(ac_ex_db.id)
-    LOG.info('[%s] Processing cancelation request for workflow.', wf_ac_ex_id)
+    LOG.info('[%s] Processing cancellation request for workflow.', wf_ac_ex_id)
```
You don't have to fix the cancellation typos here.  It's unrelated to this patch.

m4dcoder commented on st2common/st2common/services/workflows.py:
```diff
@@ -535,17 +535,24 @@ def request_action_execution(wf_ex_db, task_ex_db, st2_ctx, ac_ex_req, delay=Non
     runner_type_db = action_utils.get_runnertype_by_name(action_db.runner_type['name'])
 
+    # Identify action pack name
+    action_ref = task_ex_db.task_spec.get('spec').get('action')
```
There's already an action_ref at line 519.  They're not the same?

m4dcoder replied:
You did not address this yet.

jinpingh replied:
It is not the same. line 519 is get block of action data from action_execution_d_b. This is get action_ref = pack_name:action_name

Not sure about `You did not address this yet.` means. Changed `if else` block to one line.

m4dcoder replied:
The line numbers may have shifted.  But if you go down two line from `def request_action_execution(wf_ex_db, task_ex_db, st2_ctx, ac_ex_req, delay=None)`, there's already an action_ref assigned.  Is it different than the action_ref begin assigned in this line.

m4dcoder replied:
And if they are not the same, then you just overwrote action_ref that is saved to LiveActionDB below.

jinpingh replied:
Yes, it is different. Changed my variable name.

m4dcoder replied:
I just manually tested this and they are the same value.  Can you show me an example where they are different.

jinpingh replied:
My bad.  I thought it is reading ActionExecution.action. Put log message and tested.  Committed change.

m4dcoder replied:
Ok. Good you finally see it. I thought maybe I'm not clear enough.

m4dcoder commented on st2common/st2common/services/workflows.py:
```diff
@@ -535,17 +535,24 @@ def request_action_execution(wf_ex_db, task_ex_db, st2_ctx, ac_ex_req, delay=Non
 
+    # Identify action pack name
+    action_ref = task_ex_db.task_spec.get('spec').get('action')
+    if action_ref:
```
Make this if else block into a single line.  Use the inline if else.

m4dcoder commented on st2common/st2common/services/workflows.py:
```diff
@@ -535,17 +535,24 @@ def request_action_execution(wf_ex_db, task_ex_db, st2_ctx, ac_ex_req, delay=Non
             'task_name': task_ex_db.task_name,
             'task_id': task_ex_db.task_id
-        }
+        },
```
Why do we need a comma here?

jinpingh replied:
Did run `nosetests --nocapture st2common/tests/unit/test_param_utils.py`  and all passed

m4dcoder commented:
> Did run `nosetests --nocapture st2common/tests/unit/test_param_utils.py` and all passed

I assume you are responding to my questions about unit/integration tests.  Can you tell me what use case are we fixing here?  Where is the unit test and integration test that cover this use case?

jinpingh commented:
> > Did run `nosetests --nocapture st2common/tests/unit/test_param_utils.py` and all passed
> 
> I assume you are responding to my questions about unit/integration tests. Can you tell me what use case are we fixing here? Where is the unit test and integration test that cover this use case?

Tested `test_param_utils.py` and forgot `param.py` is not going to change. I am working on new  testcase(s) in `st2common/tests/unit/services/test_workflow.py`. Please let me know if this is not right place.

m4dcoder commented:
For the unit test, st2common/tests/unit/services/test_workflow.py looks like a good place to put the test.

For the integration test, I am also expecting the test at https://github.com/StackStorm/st2/blob/master/st2tests/integration/orquesta/test_wiring.py.  You will need st2 to be running on your dev system and then run `make orquesta-itests` to test.

m4dcoder commented:
Also, please remember to add an entry in the CHANGELOG.

jinpingh commented:
> For the unit test, st2common/tests/unit/services/test_workflow.py looks like a good place to put the test.
Done for unit test

jinpingh commented:
> For the integration test, I am also expecting a new test at https://github.com/StackStorm/st2/blob/master/st2tests/integration/orquesta/test_wiring.py. You will need st2 to be running on your dev system and then run `make orquesta-itests` to test.

For integration test,  from st2test framework, it setup `base_path` directory as `st2/st2tests/st2tests/fixtures/packs/` and database name to `st2-test`.

Since `orquesta` integration test is calling `st2client` to do Workflow executions,  so packs have to be installed before execution. Implemented code to installed two packs in the `st2-test` DB
After successfully installed two packs from `st2/st2tests/st2tests/fixtures/packs/`, observed that `orquesta_runner` is still looking for actions, workflow and virtuaenv from `/opt/stackstorm/packs/`

```error: [Errno 2] No such file or directory: u'/opt/stackstorm/packs/dummy_pack_20/actions/workflows/render_config_context.yaml'
traceback:   File "/home/vagrant/st2/st2actions/st2actions/container/base.py", line 127, in _do_run
    (status, result, context) = runner.run(action_params)
  File "/home/vagrant/st2/contrib/runners/orquesta_runner/orquesta_runner/orquesta_runner.py", line 86, in run
    wf_def = self.get_workflow_definition(self.entry_point)
  File "/home/vagrant/st2/contrib/runners/orquesta_runner/orquesta_runner/orquesta_runner.py", line 50, in get_workflow_definition
    with open(entry_point, 'r') as def_file:
```

Two suggestions:
1. Manually install a pack (second pack's workflow/action can be merged to examples), then run test. This only need to add a small function to `test_wiring.py`
2. Implemented code:
    a. Copy two packs to `/opt/stackstorm` directory
    b. Install these two packs to `st2` DB.
    c. Run test case.

Any suggestions?

m4dcoder commented:
I need to see the unit and integration tests in this PR to give you any real feedback.  When this PR is merged, circle ci will run `make orquesta-itests` to validate.  You can start by taking a look at  https://github.com/StackStorm/st2/blob/master/.circleci/config.yml and follow the integration test to see how the environment is setup and what options there are to install this other pack.

m4dcoder commented on st2common/tests/unit/services/test_workflow.py:
```diff
@@ -32,8 +32,10 @@
 from st2common.exceptions import action as action_exc
 from st2common.models.db import liveaction as lv_db_models
 from st2common.models.db import execution as ex_db_models
+from st2common.models.db.pack import ConfigDB
```
Please stay consistent with imports for this module.  Change to `from st2common.models.db import pack as pk_db_models` and move this line so the import is alphabetical.

m4dcoder commented on st2common/tests/unit/services/test_workflow.py:
```diff
@@ -32,8 +32,10 @@
+from st2common.models.db.pack import ConfigDB
 from st2common.persistence import execution as ex_db_access
 from st2common.persistence import workflow as wf_db_access
+from st2common.persistence.pack import Config
```
Since the pack module is imported above, this import is not required.

jinpingh replied:
Have to have it in order to save it to DB. Later when line 569 `param_utils.render_live_params` is called, it read config information from DB.

m4dcoder replied:
Ok. I see.  I misread.  You have to change the import here to `from st2common.persistence import pack as pk_db_access`.

jinpingh replied:
Yes.

m4dcoder commented on st2common/tests/unit/services/test_workflow.py:
```diff
@@ -44,8 +46,16 @@
+
+PACK_20 = 'dummy_pack_20'
+PACK_20_PATH = st2tests.fixturesloader.get_fixtures_packs_base_path() + '/' + PACK_20
+
```
Can we use the pack `orquesta_tests` as one of the pack here?  Also, why dummy_pack_7 and dummy_pack_20?  We can't use the other dummy_packs like dummy_pack_1?

jinpingh replied:
Yes,  we can move dummy_pack_20 to orquesta_test2. 
The reason I used two dummy pack because I was concerning about integration test. 
When install dummy_pack_1, there is a error message `Failed to register policies: Failed to register policy "/opt/stackstorm/packs/dummy_pack_1/policies/policy_2.yaml" from pack "dummy_pack_1": Referenced policy_type "action.mock_policy_error" doesnt exist` 

I almost ran all dummy packs and found these two.

jinpingh commented:
> I need to see the unit and integration tests in this PR to give you any real feedback. When this PR is merged, circle ci will run `make orquesta-itests` to validate. You can start by taking a look at https://github.com/StackStorm/st2/blob/master/.circleci/config.yml and follow the integration test to see how the environment is setup and what options there are to install this other pack.

m4dcoder commented on st2common/st2common/services/workflows.py:
```diff
@@ -1110,4 +1114,4 @@ def update_execution_records(wf_ex_db, conductor, update_lv_ac_on_states=None,
     if status_changed and wf_lv_ac_db.status in ac_const.LIVEACTION_COMPLETED_STATES:
         LOG.info('[%s] Workflow action execution is completed and invoking post run.', wf_ac_ex_id)
-        runners_utils.invoke_post_run(wf_lv_ac_db)
+        runners_utils.invoke_post_run(wf_lv_ac_db)
```
Remove the carriage return here.

jinpingh commented:
> I need to see the unit and integration tests in this PR to give you any real feedback. When this PR is merged, circle ci will run `make orquesta-itests` to validate. You can start by taking a look at https://github.com/StackStorm/st2/blob/master/.circleci/config.yml and follow the integration test to see how the environment is setup and what options there are to install this other pack.

From `st2cd`  https://github.com/StackStorm/st2cd/blob/master/actions/setup_e2e_tests.sh, where `examples` pack is installed. We can add one more pack here:
```sudo cp -R /usr/share/doc/st2/examples /opt/stackstorm/packs/
st2 run packs.setup_virtualenv packs=examples,tests,asserts,fixtures,webui
st2ctl reload --register-all
```

Will find out how file is located at `/usr/share/doc/st2`.  Looking at https://github.com/StackStorm/st2/blob/master/.circleci/config.yml

m4dcoder commented on st2tests/integration/orquesta/test_wiring.py:
```diff
@@ -144,3 +147,19 @@ def test_output_on_error(self):
         self.assertDictEqual(ex.result, expected_result)
+
+    def test_config_context_renders(self):
+        packs = ["orquesta_tests", "dummy_pack_7"]
```
These are packs for unit tests.  We have separate ones for integration tests.  I use the examples pack.  But there are other packs that st2 CI uses.  Looking at how the st2 environment is setup in CI and reusing existing test packs and installing those test packs on setup is simpler than writing an entire InstallFixturesPacks class here.

jinpingh replied:
> These are packs for unit tests. We have separate ones for integration tests. I use the examples pack. But there are other packs that st2 CI uses. Looking at how the st2 environment is setup in CI and reusing existing test packs and installing those test packs on setup is simpler than writing an entire InstallFixturesPacks class here.

Submitted code. Want to make sure I am in the right track. Did look at existing pack to replace dummy_pack_7 and still did not get the right one. I think we should create a new pack in the case integration test like this that needs too packs  with complicated setup. Thanks!

m4dcoder commented on st2tests/st2tests/base.py:
```diff
@@ -465,6 +469,69 @@ def setUp(self):
             self._register_pack_configs()
 
 
+class InstallFixturesPacks(DbTestCase):
```
Please look for a simpler alternative first thru the environment setup in circleci.  May be simpler than writing a brand new class to do pack and venv setup here.

m4dcoder commented on tools/launchdev.sh:
```diff
@@ -212,6 +212,8 @@ function st2start(){
     if [ "$copy_examples" = true ]; then
         echo "Copying examples from ./contrib/examples to $PACKS_BASE_DIR"
         cp -Rp ./contrib/examples $PACKS_BASE_DIR
+        cp -Rp ./st2tests/st2tests/fixtures/packs/dummy_pack_7 $PACKS_BASE_DIR
```
We have a `test` pack for our integration tests @ https://github.com/StackStorm/st2tests/tree/master/packs/tests.  Can we use this instead of dummy_pack_7?

jinpingh replied:
Only issue by using this one is we have to put ` st2tests/st2tests/fixtures/packs/configs/dummy_pack_7.yaml` file under `st2tests/tests/configs/test.yaml`

m4dcoder replied:
Why is this an issue adding a new config schema in the `tests` pack?

jinpingh replied:
Thought it is not normal way to place pack config file. 
Did change dummy_7_pack to `tests` and please review it. See https://github.com/StackStorm/st2/pull/4570/commits/c81f04270d3f7f442652c1adf84e38287968ef95
Thanks!

m4dcoder commented on st2tests/tests/configs/tests.yaml.yaml:
```diff
@@ -0,0 +1,2 @@
+---
```
Shouldn't this file be `st2tests/tests/configs/tests.yaml`?  There's two yamls in the file name.

m4dcoder commented on st2tests/tests/config.schema.yaml:
```diff
@@ -0,0 +1,5 @@
+config_item_one:
+  description: "Item use to test config context."
+  type: "string"
+  required: true
```
What it work if you add a default value here?  Then you wouldn't need to add `st2tests/tests/configs/tests.yaml`.

m4dcoder replied:
The tests pack I'm asking you to modify is located @ https://github.com/StackStorm/st2tests/tree/master/packs/tests.  It is a separate repo.  You're adding files in the wrong place.

m4dcoder commented on st2tests/tests/actions/render.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+name: render
```
Can you name this better so we know what we are testing and what we are rendering?

m4dcoder commented on CHANGELOG.rst:
```diff
@@ -33,6 +33,8 @@ Fixed
 * Fix orquesta workflow bug where context variables are being overwritten on task join.
   (bug fix) StackStorm/orquesta#112
+* Fix `config_context` renders against incorrect pack 
+  (bug fix) StackStorm/st2#4570  
```
Can you change this to "Fix rendering of `config_context` in orquesta task that references action in different pack (bug fix) #4570"?

m4dcoder commented on st2common/st2common/services/workflows.py:
```diff
@@ -541,9 +541,12 @@ def request_action_execution(wf_ex_db, task_ex_db, st2_ctx, ac_ex_req, delay=Non
     runner_type_db = action_utils.get_runnertype_by_name(action_db.runner_type['name'])
 
+    # Identify action pack name
+    pack_name = action_ref.split('.')[0] if action_ref else st2_ctx.get('pack')
```
Orquesta task can be action less. Example at https://github.com/StackStorm/st2/blob/master/contrib/examples/actions/workflows/tests/orquesta-test-action-less-tasks.yaml#L12.  Please add a use case for this line that shows it can handle task that is action less.

jinpingh replied:
Tested of action less workflow `examples.orquesta-test-action-less-tasks`  and it went well. Not sure about `add a use case`, is to add comment or unit test case?

m4dcoder replied:
Ok. Then we're good.  Just want to make sure this use case is covered.  Thank you for your diligence.  It's appreciated.

m4dcoder commented on st2common/st2common/services/workflows.py:
```diff
@@ -541,9 +541,12 @@ def request_action_execution(wf_ex_db, task_ex_db, st2_ctx, ac_ex_req, delay=Non
     # Set context for the action execution.
     ac_ex_ctx = {
-        'pack': st2_ctx.get('pack'),
+        'pack': pack_name,
```
Please default to st2_ctx.get('pack') when task is action less.

jinpingh replied:
` pack_name = action_ref.split('.')[0] if action_ref else st2_ctx.get('pack')`
With above line,  for action less, pack will be default to `st2_ctx.get('pack')` , not sure what need to be changed here. Please give more information.

m4dcoder replied:
We're good here then.  See comment above. Thanks for following up.

m4dcoder commented on tools/launchdev.sh:
```diff
@@ -212,6 +212,8 @@ function st2start(){
     if [ "$copy_examples" = true ]; then
```
Can you rename `copy_examples` to `copy_test_packs`?

m4dcoder commented on tools/launchdev.sh:
```diff
@@ -413,6 +415,10 @@ function st2start(){
     fi
 
+    if [ "$copy_examples" = true ]; then
+        st2 run packs.setup_virtualenv packs=tests
```
I believe this will fail when auth is enabled in the dev config.  Can you test this scenario and make sure the launch script handle this gracefully and warn users that they have to setup the virtualenv on their own.

m4dcoder commented on st2tests/tests/actions/render.yaml:
```diff
@@ -0,0 +1,12 @@
+---
```
The tests pack I'm asking you to modify is located @ https://github.com/StackStorm/st2tests/tree/master/packs/tests.  It is a separate repo.  You're adding files in the wrong place.

jinpingh replied:
Sorry about this, did not realize there is a different repo for `st2tests`.  On cicd for integration test, does `tests` pack installed on the same directory as https://github.com/StackStorm/st2/tree/master/st2tests/tests? Because it needs to be copied to `/opt/stackstrom/packs`, need to know the source location.

Since i don't understand how https://github.com/StackStorm/st2tests/tree/master/packs/tests is used, for earlier unit test submission, `dummy_pack_7` is used for testing, should we change it to the same `tests` pack?

m4dcoder commented on st2tests/tests/actions/render.py:
```diff
@@ -0,0 +1,7 @@
+from st2common.runners.base_action import Action
```
The tests pack I'm asking you to modify is located @ https://github.com/StackStorm/st2tests/tree/master/packs/tests.  It is a separate repo.  You're adding files in the wrong place.

m4dcoder commented on st2tests/tests/config.schema.yaml:
```diff
@@ -0,0 +1,5 @@
+---
```
The tests pack I'm asking you to modify is located @ https://github.com/StackStorm/st2tests/tree/master/packs/tests.  It is a separate repo.  You're adding files in the wrong place.

m4dcoder commented:
@jinpingh I just want to let you know that you are doing great here.  We are near the finish line for this PR.  There are a number of things that is new here (registering pack besides examples pack for integration test) that even if I am the developer working on this, it would involve me doing exploring and trial and error.  So I expected back and forth with this PR. The feedbacks on the review has more to do with helping you gain knowledge and get into the habit of writing unit tests and integration tests then anything else.  Thank you for your diligence. So take a breather and continue the good work.

jinpingh commented on tools/launchdev.sh:
```diff
@@ -209,11 +209,10 @@ function st2start(){
         cp -Rp ./contrib/examples $PACKS_BASE_DIR
-        cp -Rp ./st2tests/tests $PACKS_BASE_DIR
-        cp -p ./st2tests/tests/configs/tests.yaml $CONFIG_BASE_DIR
+        cp -Rp ./st2tests/packs/tests $PACKS_BASE_DIR
```
Please let me know if this is the right path `./st2tests/packs/tests`

m4dcoder replied:
The `tests` pack is in a completely different repo https://github.com/StackStorm/st2tests/. You have to git clone https://github.com/StackStorm/st2tests.git to a temporary location and then copy the directory `./st2tests/packs/tests/` from the temporary location to /opt/stackstorm/packs.

m4dcoder commented on tools/launchdev.sh:
```diff
@@ -415,8 +414,24 @@ function st2start(){
+                    break
+                fi
+            fi
+        done < "$ST2_CONF"
```
Instead of this while loop, can we just run `st2 run packs.setup_virtualenv packs=tests` and if it returns `ERROR: 401 Client Error: Unauthorized`, you echo the warning?

jinpingh replied:
I thought `just run st2 run packs.setup_virtualenv packs=tests and if it returns ERROR: 401 Client Error: Unauthorized` is easy way but not graceful way.

m4dcoder replied:
It's ok with me.

m4dcoder replied:
Are you going to change this?

jinpingh replied:
> Are you going to change this?

I am not going to change, but if this is too much, i can change it.

m4dcoder replied:
Yeah, I prefer simplicity here.  If it returns an auth error, then warn users.  This is a dev launch script, don't need to go thru the hassle to parse for the auth lines in the config file.

Kami commented:
Just adding this info here for completeness. It looks like this is related to - https://github.com/StackStorm/st2/pull/4531.

There I fixed the original issue, but I didn't correctly handle child workflow action which this PR fixes :+1:

m4dcoder commented:
@jinpingh We are almost done here.  So after these minor changes, get everything green before asking for review again.

m4dcoder commented on tools/launchdev.sh:
```diff
@@ -413,6 +424,13 @@ function st2start(){
+    if [ "$copy_test_packs" = true ]; then
+        st2 run packs.setup_virtualenv packs=tests
+        if [ $? != 0 ]; then
+            echo "Warning: Please setup virtualenv for pack \"tests\" before run integration test"
```
Please change the message to `Unable to setup virtualenv for the \"tests\" pack. Please setup virtualenv for the \"tests\" pack before running integration tests.`

Kami commented:
Just a heads up - I didn't do extensive end to end testing (e.g. multiple levels of nested workflows and executions), but I tested it with example packs from https://github.com/bigmstone/packConfig and it appears to be working fine.

Kami commented:
Since it got merged in time, I will also cherry pick this changes in a branch for v2.10.4 release :+1:
